### DISCUSSION TITLE
Testing a serverless Node API locally within an Angular project

### DISCUSSION THREAD

**Discussion by dotNetkow** [vercel/vercel#6578] [Help] [answered]
I have an Angular project and have added an "api" folder along with a TypeScript file, `start.ts`. If I deploy the project to Vercel, the API works at `<URL>/api/start`.

However, how do I get `vercel dev` to work so I can locally test out the API? Since this is an Angular project originally, the Development Command is set to `ng serve` - but that doesn't build/deploy the API since it's not part of the Angular app. So, any API requests 404 because it doesn't exist. 

Setting the project Settings to "Other" then overriding Dev Command and setting it to blank also fails, with error:
```
The specified command ("start") is invalid. For a list of available options,
run "ng help". Did you mean "test"?
```

**Comment by leerob** [comment]
Running locally with `vercel dev` should serve your Angular application, as well as the API route in the `/api` folder. Could you share more information if this isn't working as expected?

**Comment by dotNetkow** [comment]
hey @leerob thank you for the quick response! Here's the app on GitHub and [the API](https://github.com/ionic-team/ionifits/blob/payments/api/session/start.ts). I'm just trying to use the Vercel-provided example for now. When it's deployed on Vercel, it works. for example, make a POST request here: https://ionifits-qxoo16c9t-ionic1.vercel.app/api/session/start.

When I run vercel dev locally, and using Postman make that same request (albeit on localhost like http:localhost:4000/api/session/start) it gives me a 404. 

Project settings:
Build command: ng build --prod
output directory: www (this is an Ionic Angular hybrid app)
install command: none
Development command: ng serve (also tried blank, no dice)

**Comment by dotNetkow** [comment, marked as answer]
I figured it out! I had a hunch that a Next app would work locally, since Vercel created and maintains it. Sure enough, that was true. After running `vercel dev`, the console showed: `next dev --port $PORT`.

The port flag was missing. That's the secret sauce that makes the API work. Turns out this is documented on the [Build Step page](https://vercel.com/docs/build-step?query=port#development-command). It would be helpful to also include this info on the main Serverless Functions page, [Local Development section](https://vercel.com/docs/serverless-functions/introduction#local-development). It doesn't mention the Port requirement at all. I don't see a way to submit a docs PR, so I left feedback via the smiley face comment widget.

So, in order for local Angular debugging to work:

```
ng serve --port $PORT
```

Both the Angular app and the Serverless API are locally available for testing.

**Reply by leerob** [reply]
Are you using the latest version of the Vercel CLI? It should be setting this for you

- https://github.com/vercel/vercel/blob/main/packages/frameworks/src/frameworks.ts#L731
- https://github.com/vercel/vercel/blob/main/packages/frameworks/src/frameworks.ts#L739

**Reply by leerob** [reply]
I will update the docs to make this more clear, thank you!

**Reply by dotNetkow** [reply]
Thanks, I think I got it - this is an Ionic Angular project! that config [only specifies `ng serve` for dev command](https://github.com/vercel/vercel/blob/main/packages/frameworks/src/frameworks.ts#L685) unlike regular Angular which includes the PORT. I'll PR it...

**Reply by dotNetkow** [reply]
OK, opened a PR! https://github.com/vercel/vercel/pull/6658.
